lib/DxilRootSignature/DxilRootSignature.cpp

```diff
@@ -190,6 +190,44 @@ void DeleteRootSignature(const DxilVersionedRootSignatureDesc * pRootSignature)
   delete pRootSignature;
 }
 
+template <typename RootSigDesc, typename RootParameterDesc>
+void ExtendRootSig(RootSigDesc &rootSigDesc, 
+                   uint32_t ShaderRegister,
+                   uint32_t RegisterSpace) {
+  auto *existingParams = rootSigDesc.pParameters;
+  auto *newParams = new RootParameterDesc[rootSigDesc.NumParameters + 1];
+  memcpy(newParams, existingParams, rootSigDesc.NumParameters * sizeof(RootParameterDesc));
+  if (existingParams != nullptr) {
+    delete[] existingParams;
+  }
+  rootSigDesc.pParameters = newParams;
+  rootSigDesc.pParameters[rootSigDesc.NumParameters].ParameterType = DxilRootParameterType::UAV;
+  rootSigDesc.pParameters[rootSigDesc.NumParameters].Descriptor.RegisterSpace = RegisterSpace;
+  rootSigDesc.pParameters[rootSigDesc.NumParameters].Descriptor.ShaderRegister = ShaderRegister;
+  rootSigDesc.pParameters[rootSigDesc.NumParameters].ShaderVisibility = DxilShaderVisibility::All;
+  rootSigDesc.NumParameters++;
+}
+
+void AddDescriptorParamter(const DxilVersionedRootSignatureDesc *pRootSignature,
+                           uint32_t ShaderRegister, 
+                           uint32_t RegisterSpace) {
+
+  auto *rs = const_cast<DxilVersionedRootSignatureDesc *>(pRootSignature);
+  switch (pRootSignature->Version) {
+  case DxilRootSignatureVersion::Version_1_0:
+    ExtendRootSig<DxilRootSignatureDesc, DxilRootParameter>(rs->Desc_1_0,
+                                                            ShaderRegister, RegisterSpace);
+    break;
+  case DxilRootSignatureVersion::Version_1_1:
+    ExtendRootSig<DxilRootSignatureDesc1, DxilRootParameter1>(rs->Desc_1_1,
+                                                            ShaderRegister, RegisterSpace);
+    rs->Desc_1_1.pParameters[rs->Desc_1_1.NumParameters - 1].Descriptor.Flags =
+        hlsl::DxilRootDescriptorFlags::None;
+    break;
+  }
```

Comment:
If this can return rs, then I think the call to AddUAVParamterToRootSignature() can be replaces with:

```c++
auto rs = AddDescriptorParamter();
auto extendedRootSig = SerializeRootSignatureToVector(rs);
```

Reply:
DxilRootSingature.cpp's public surface uses "const" in all its DxilVersionedRootSignatureDesc pointers. Presumably it's doing this to enforce the contract that nobody outside of DxilRootSignature.cpp owns the memory allocation, nor is allowed to modify the structure. Returning rs (the non-const) would mess with this convention.

Reply:
As noted below, I'll bow to the consensus and return a non-const pointer to PIX. THis helps clean this up a bit.